python/ray/ml/train/integrations/huggingface/huggingface_trainer.py

```diff
@@ -146,52 +174,391 @@ def trainer_init_per_worker(train_dataset, eval_dataset, **config):
+            return
+        self.last_step = state.global_step
+        report = {**logs, "step": state.global_step, "epoch": state.epoch}
+        if control.should_save:
+            self.delayed_report = report
+        else:
+            train.report(**report)
+
+    def on_save(self, args, state, control, **kwargs):
+        checkpoint_path = Path(
+            transformers.trainer.get_last_checkpoint(args.output_dir)
+        ).absolute()
+        if checkpoint_path:
+            train.save_checkpoint(
+                **{
+                    NODE_IP_KEY: get_node_ip_address(),
+                    CHECKPOINT_PATH_ON_NODE_KEY: str(checkpoint_path),
+                }
+            )
+        if self.delayed_report:
+            train.report(**self.delayed_report)
+            self.delayed_report = None
+        gc.collect()
+
+
+def _process_dataset_for_hf(
+    dataset: Dataset, feature_columns: Dict[str, List[str]], batch_size: int = 1
+) -> IterableDataset:
+    """Converts a Ray Dataset into a HF-compatible Torch Dataset."""
+    torch_dataset = dataset.to_torch(
+        batch_size=batch_size,
+        feature_columns=feature_columns,
+        label_column="labels",
+        unsqueeze_label_tensor=False,
+        unsqueeze_feature_tensors=False,
+    )
+    try:
+        count = dataset.count()
+    except ValueError:
+        # pipeline case
+        count = None
+    if count:
+        torch_dataset = _HFIterableDatasetWithLen(torch_dataset, count)
+    return torch_dataset
+
+
+# TODO(team-ml): Refactor checkpoint management along with Tune.
+class _DataParallelSyncingCheckpointManager(TuneCheckpointManager):
+    """Same as _DataParallelCheckpointManager, but syncs the dir instead
+    of serializing it."""
+
+    def add_tune_checkpoint_id(self, path: str):
+        # Store the checkpoint_id in the file so that the Tune trial can be
+        # resumed after failure or cancellation.
+        with open(Path(path).joinpath(TUNE_CHECKPOINT_ID), "w") as f:
+            f.write(str(self._latest_checkpoint_id))
+
+    def on_init(self, preprocessor: Preprocessor):
+        self.preprocessor = preprocessor
+        super(_DataParallelSyncingCheckpointManager, self).on_init()
+
+    def write_checkpoint(self, checkpoint: Dict):
+        # If inside a Tune Trainable, then checkpoint with Tune.
+        with tune.checkpoint_dir(step=self._latest_checkpoint_id) as checkpoint_dir:
+            source_ip = checkpoint[NODE_IP_KEY]
+            source_path = checkpoint[CHECKPOINT_PATH_ON_NODE_KEY]
+            target_ip = get_node_ip_address()
+            sync_dir_between_nodes(
+                source_ip=source_ip,
+                source_path=source_path,
+                target_ip=target_ip,
+                target_path=checkpoint_dir,
+                return_futures=False,
+                max_size_bytes=None,
+            )
+            if source_ip == target_ip:
+                shutil.rmtree(source_path, ignore_errors=True)
```

Comment:
Maybe we can clean up here
```suggestion
            if source_ip == target_ip:
                shutil.rmtree(source_path, ignore_errors=True)
            else:
                delete_on_node(node_ip=source_ip, path=source_path)
```

By the way, if source == target, should we just move the checkpoint dir instead of calling the sync function?